tests/test_aspa_vector.c

```diff
@@ -0,0 +1,163 @@
+/*
+* This file is part of RTRlib.
+*
+* This file is subject to the terms and conditions of the MIT license.
+* See the file LICENSE in the top level directory for more details.
+*
+* Website; http://rtrlib.realmv6.org/
+*/
+
+#include "rtrlib/lib/vector.h"
+
+#include <assert.h>
+
+static void test_create_vector()
+{
+	struct aspa_vector *vector;
+	assert(create_vector(&vector) == 0);
+	assert(vector->data != NULL);
+	assert(vector->size == 0);
+	assert(vector->capacity == 100);
+};
+
+static void generate_fake_aspas_record(uint32_t cas, uint32_t random_number, size_t amount,
+				       struct aspas_record **record)
+{
+	struct aspas_record *new_record = (struct aspas_record *)malloc(sizeof(struct aspas_record));
+	new_record->customer_as_number = cas;
+	uint32_t *provider_as_numbers = (uint32_t *)malloc(sizeof(uint32_t) * amount);
+
+	for (size_t i = 0; i < amount; i++) {
+		provider_as_numbers[i] = random_number + i;
+	}
+
+	new_record->number_of_provider_as = amount;
+	new_record->provider_as_numbers = provider_as_numbers;
+
+	*record = new_record;
+}
+
+static void test_add_element()
+{
+	struct aspa_vector *vector;
+	assert(create_vector(&vector) == 0);
+
+	struct aspas_record *record;
+	generate_fake_aspas_record(42, 300, 3, &record);
+	add_element_vector(vector, *record);
+
+	assert(vector->data[0].customer_as_number == 42);
+	assert(vector->data[0].number_of_provider_as == 3);
+	assert(vector->data[0].provider_as_numbers[0] == 300);
+	assert(vector->data[0].provider_as_numbers[1] == 300 + 1);
+	assert(vector->data[0].provider_as_numbers[2] == 300 + 2);
+}
+
+static void test_relocate()
+{
+	struct aspa_vector *vector;
+	assert(create_vector(&vector) == 0);
+	vector->capacity = 2;
+	struct aspas_record *old_pointer = vector->data;
+
+	// this is on purpose wrongly ordered so we check that it does the sort properly
+	struct aspas_record *record_2;
+	generate_fake_aspas_record(1, 100, 3, &record_2);
+	add_element_vector(vector, *record_2);
+
+	struct aspas_record *record_1;
+	generate_fake_aspas_record(2, 200, 3, &record_1);
+	add_element_vector(vector, *record_1);
+
+	struct aspas_record *record_4;
+	generate_fake_aspas_record(3, 300, 3, &record_4);
+	add_element_vector(vector, *record_4);
+
+	struct aspas_record *record_3;
+	generate_fake_aspas_record(4, 400, 3, &record_3);
+	add_element_vector(vector, *record_3);
+
+	assert(old_pointer != vector->data); // new pointer because relocated
+	assert(vector->capacity == 102);
```

Comment:
I wouldn't test for capacity here as it's an implementation detail not relevant to the user

Reply:
but its relevant for the functionality :D

Reply:
maybe check if the capacity is large enough (>) instead of at just the right capacity (==)